slurm_openstack_tools/reboot.py

```diff
@@ -1,160 +1,151 @@
+    # Fetch the server object once
+    server = find_server(conn, server_id)
+    if not server:
+        return
 
-    Requires:
-    - Python 3 with openstacksdk module
-    - The node's Openstack ID to have been set by cloud init in
-      `/var/lib/cloud/data/instance-id`
-    - An application credential:
-        - with at least POST rights to /v3/servers/{server_id}/action
-        - available via a clouds.yaml file containing only one cloud
-    """ # noqa E501
-    server_uuid = get_openstack_server_id()
-    if not server_uuid:
-        logger.info("rebooting non openstack server")
-        do_reboot()
+    current_image_id = server.image['id'] if 'image' in server and server.image else None
+    if current_image_id != target_image_id:
+        logger.info(f"Node {node} requires rebuild: current image {current_image_id}, target image {target_image_id}")
+        rebuild_node(conn, server, target_image_id)
     else:
-        reason = get_reboot_reason()
-        if not reason.startswith("rebuild"):
-            logger.info("rebooting openstack server, locally")
-            do_reboot()
-        else:
-            logger.info("rebuilding openstack server")
-            rebuild_openstack_server(server_uuid, reason)
+        logger.info(f"Node {node} is already using the target image, performing reboot...")
+        reboot_node(conn, server)
+
+
+def rebuild_node(conn, server, target_image_id):
+    """
+    Rebuild the node with the target image using the server object.
+    """
+    image = conn.image.get_image(target_image_id)
+    if not image:
+        logger.error(f"Target image {target_image_id} not found in OpenStack")
+        return False
+
+    logger.info(f"Rebuilding server {server.id} with image {target_image_id}")
+    conn.rebuild_server(server, image)
+    return True
+
+
+def reboot_node(conn, server, reboot_type="SOFT"):
+    """
+    Reboot the node using the server object (default: soft reboot).
+    """
+    logger.info(f"Rebooting server {server.id} with {reboot_type.lower()} reboot")
+    conn.compute.reboot_server(server, reboot_type=reboot_type)
+    return True
 
 
 def main():
+    """
+    Main function to process nodes from the Slurm-provided hostlist.
+    """
+    if len(sys.argv) < 2:
+        logger.error("Usage: <script> <hostlist>")
+        sys.exit(1)
+
+    hostlist = sys.argv[1].split(",")
+
     try:
-        rebuild_or_reboot()
-    except BaseException:
-        logger.exception('Exception in rebuild_or_reboot():')
-        raise
+        conn = openstack.connection.from_config()
+        logger.debug("OpenStack connection established")
+    except Exception as e:
+        logger.error(f"Failed to establish OpenStack connection: {e}")
+        sys.exit(1)
+
+    for node in hostlist:
+        logger.debug(f"Processing node: {node}")
+        try:
+            process_node(conn, node)
+        except Exception as e:
+            logger.error(f"Failed to process node {node}: {e}")
```

Comment:
Ah, so I think we need to work out if there were any errors, and if they were our script needs to exist with an error code, so Slurm doesn't get confused.